Support KPR with IPv6 underlay

pchaigno opened the pull request:
This pull request adds support for KPR when running with an IPv6 underlay. The first commit introduces an IPv6 version of the encapsulation function. The second commit performs some refactoring to ease readability of subsequent commits. The third commit removes a function argument that became unnecessary thanks to the previous commit. The fourth commit adds the support in the datapath. Fifth and last commit enable this in the agent and cover it in the CI.

```release-note
Add support for kube-proxy replacement with IPv6 underlay
```

pchaigno committed:
bpf: Support encapsulation with options for IPv6

When using GENEVE tunneling, we can include options in the tunnel
header. That is used by the NodePort logic in some cases. This commit
adds support for doing this over an IPv6 underlay, by extending
ctx_set_encap_info6 and introducing __encap_with_nodeid_opt6 to match
their IPv4 counterparts.

There are no functional changes in this commit because
__encap_with_nodeid_opt6 isn't used yet.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf/nodeport: Pass whole ipcache value to encap functions

Encap functions nodeport_add_tunnel_encap{,_opt} currently take the
IPv4 tunnel endpoint address as anargument. These functions will need
to also support IPv6 addresses so let's first pass, as an argument, a
pointer to the whole ipcache value (i.e., struct remote_endpoint_info)
instead of the tunnel endpoint. It will be easier in subsequent commits
to retrieve either the IPv4 or IPv6 tunnel endpoint from the ipcache
value.

There are no functional changes in this commit.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf/nodeport: Support IPv6 underlay for tunneling

In tunneling mode with BPF NodePort, when DNATed packets are forwarded
to a remote backend, they need to be encapsulated. This commit
implements support for an IPv6 underlay in that case.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
workflows/e2e: Cover IPv6-only + VXLAN

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
daemon, workflows: Support KPR + IPv6 underlay

Previous commits added support for IPv6 underlays with BPF NodePort so
we can now allow users to enable those options together. This commit
also switches the IPv6 underlay test case in end-to-end tests to use KPR
and egress gateway.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -125,9 +131,14 @@ nodeport_add_tunnel_encap_opt(struct __ctx_buff *ctx, __u32 src_ip, __be16 src_p
-				       opt, opt_len, ct_reason, monitor, ifindex);
+	if (info->flag_ipv6_tunnel_ep)
+		return __encap_with_nodeid_opt6(ctx, &info->tunnel_endpoint.ip6,
+						src_sec_identity, info->sec_identity,
```
```suggestion
						src_sec_identity, dst_sec_identity,
```

until we're confident that all callers pass the identical identity in the `remote_endpoint_info`, and can then also use it in the IPv4 path?

pchaigno replied:
I went the other way around and completely remove the `dst_sec_identity` argument.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1147,7 +1148,7 @@ int tail_nodeport_nat_egress_ipv6(struct __ctx_buff *ctx)
 	info = lookup_ip6_remote_endpoint(dst, 0);
-	if (info && info->tunnel_endpoint.ip4 != 0 && !info->flag_skip_tunnel) {
+	if (info && info->flag_has_tunnel_ep && !info->flag_skip_tunnel) {
 		tunnel_endpoint = info->tunnel_endpoint.ip4;
```
I think we also want to update the condition [here](https://github.com/cilium/cilium/blob/12ef4ab6240651e555957611f61dab2e623e3f82/bpf/lib/nodeport.h#L1178) now, to check for `info->flag_has_tunnel_ep && !info->flag_skip_tunnel`.

pchaigno replied:
Fixed.

julianwiedmann reviewed (commented):
two minor comments, except that looks good! Adding the XDP support will be fun ...

pchaigno committed:
bpf/nodeport: Remove unused dst_sec_identity arg

Since the previous commit passes the whole ipcache value into
nodeport_add_tunnel_encap{,opt}, we don't need the dst_sec_identity
argument anymore. We can retrieve its value from the ipcache value.

This commit therefore removes the dst_sec_identity argument from the
nodeport_add_tunnel_encap{,opt} functions.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

julianwiedmann reviewed (approved):
looks good, thank you!
